client/src/unifyfs.c

```diff
@@ -1109,21 +1149,31 @@ int unifyfs_gfid_read_reqs(read_req_t* in_reqs, int in_count)
      * are missed
      */
 
+    LOGDBG("%d read request(s) submitted to the server (%d files)",
+           count, gfid_count);
+
     /* spin waiting for read data to come back from the server,
      * we process it in batches as it comes in, eventually the
-     * server will tell us it's sent us everything it can */
-    int done = 0;
-    while (!done) {
+     * server will tell us it's sent us everything it can.
+     * NOTE: for mread, the completion (done) is reported per file (gfid). */
+    int gfid_done = 0;
+    while (gfid_done < gfid_count) {
         int tmp_rc = delegator_wait();
         if (tmp_rc != UNIFYFS_SUCCESS) {
             rc = UNIFYFS_FAILURE;
-            done = 1;
+            break;
         } else {
+            int done = 0;
             tmp_rc = process_read_data(read_reqs, count, &done);
```

Comment:
I was hoping that ```read_reqs``` could contain requests that could span over multiple files and then the server would signal its down with all requests.

I guess the server is only responding with one file at a time?  Can we fix things on the server side instead?  It sounds like the server is telling the client it is done too early.

Reply:
@adammoody It seems like the server doesn't have a data structure to represent a `lio_listio` request consisting of multiple requests with multiple files (file descriptors). The server only tracks the request per gfid (`rm_cmd_mread` and `create_gfid_chunk_read`):

https://github.com/LLNL/UnifyFS/blob/b4ea3eeceac1c9bf4dffc95ac7479ae3003bd3e3/server/src/unifyfs_request_manager.c#L1313

I will take a 2nd look as you suggested, but it may cause more changes.

Reply:
@adammoody in the new commit, I've changed the server data structures so it can send a single complete message per client request.